Adds new diagrams for timestamps (#223)
amzn/ion-docs: src/binary-encoding.adoc

|===
_{asterisk} Serialized size in bytes does not include the opcode._

> The body of short-form timestamps are encoded as a `FixedUInt` of the size specified by the opcode.  This integer is

Reviewer: ```suggestion
The body of a short-form timestamp is encoded as a `FixedUInt` of the size specified by the opcode.  This integer is
```

> The body of short-form timestamps are encoded as a `FixedUInt` of the specified size of the op-code.  This integer is
> then partitioned in bit-fields for the timestamp's subfields.  Note that endian does not apply here because the
> bit-fields are defined over the body interpreted as an _integer_.
> 

Reviewer 2: ```suggestion
The body of short-form timestamps are encoded as a `FixedUInt` of the size specified by the opcode.  This integer is
then partitioned into bit-fields representing the timestamp's subfields.  Note that endianness does not apply here because the
bit-fields are defined over the body interpreted as an _integer_.

```

I think the "here" in "does not apply here" could be spelled out a bit more.

30 (ns) +
| Fractional second

| `.`
| n/a
| Unused
|===

> We will denote the timestamp encoding as follows with each byte ordered from vertically from top to bottom.  The

Reviewer 2: ```suggestion
We will denote the timestamp encoding as follows with each byte ordered vertically from top to bottom.  The
```

respective bits are denoted using the letter codes defined in the table above.

[%unbreakable,source]
----
          7       0 <--- bit position
          |       |
         +=========+
byte 0   |  0xNN   | <-- hex notation for constants like opcodes

>          +=========+ <-- boundary between some encoding primitive (e.g., opcode/`FlexUInt`)

Reviewer 2: > between some encoding primitive (...) 

The word "between" made me expect a second clause. ("Between some encoding primitive and ________").

Author: I will edit this as:

> boundary between encoding primitives (e.g., opcode/`FlexUInt`)

     1   |nnnn:nnnn| <-- bits denoted with a `:` as a delimeter to aid in reading
         +---------+ <-- octet boundary within an encoding primitive
         ...
         +---------+
     N   |nnnn:nnnn|
         +=========+
----

> The bytes are read from top to down (least significant to most significant), while the bits within each byte should be

Reviewer 2: ```suggestion
The bytes are read from top to bottom (least significant to most significant), while the bits within each byte should be
```
or
```suggestion
The bytes are read from the top down (least significant to most significant), while the bits within each byte should be
```

* *12 bits for the offset*, which counts the number of minutes (not quarter-hours) from -1440
(that is: `-24:00`). An offset value of `0b111111111111` indicates an unknown offset.

Similar to short-form timestamps, with the exception of representing the fractional seconds, the components of the
timestamp are encoded as bit-fields on a `FixedUInt` that corresponds to the length that followed the opcode.

> If the timestamp's overall length is greater than or equal to `8`, the `FixedUInt` part of the timestamp is `8` bytes

Reviewer: ```suggestion
If the timestamp's overall length is greater than or equal to `8`, the `FixedUInt` part of the timestamp is `7` bytes
```

I think this is what you're trying to say? It takes 7 bytes to encode a long-form timestamp through seconds, then the FlexUInt + FixedUInt.

Author: Great catch!

> and the remainder of the length is used to encode fractional seconds. The fractional seconds are encoded as a

Reviewer: ```suggestion
and the remaining bytes are used to encode fractional seconds. The fractional seconds are encoded as a
```

`(coefficient, scale)` pair, which is _similar_ to a <<decimals, decimal>>. The primary difference is that the *scale*
represents a negative *exponent* because it is illegal for the fractional seconds value to be greater than or equal to
`1.0` or less than `0.0`. The coefficient is encoded as a `FlexUInt` (instead of `FlexInt`) to prevent the encoding of
fractional seconds less than `0.0`. The scale is encoded as a `FixedUInt` (instead of `FixedInt`) to discourage the
encoding of decimal numbers greater than `1.0`. Note that validation is still required; namely:

* A scale value of `0` is illegal, as that would result in a fractional seconds greater than `1.0` (a whole second).
* If `coefficient * 10^-scale > 1.0`, that `(coefficient, scale)` pair is illegal.

If the timestamp's length is `3`, the most significant bit in the final byte (`h`) is a flag
that indicates month (`0`) or day (`1`) precision. If the timestamp's length is greater than `3`, the (`h`) bit is
treated as the least-significant bit of the hour (`H`) bits.

.Figure {counter:figure}: Encoding of the _body_ of a long-form timestamp
[%unbreakable,source]
----
         +=========+

> byte 0   |YYYY:YYYY|

Reviewer: In the other diagrams, byte 0 began with the opcode. Should we do that here too, for consistency?

Author: In this one, it was specifically only the body.  I had added #222 before this PR to cover this because when I went through this section I realized that we should have a bunch of examples w/opcodes.
